src/api/mod.rs

```diff
@@ -403,6 +451,119 @@ impl<'a> ValueListBuilder<'a> {
     }
 }
 
+//TODO: move to its own module metadata, with related types
+fn to_meta_data(meta_hm: &HashMap<String, MetaValue>) -> Result<*mut meta_data_t, SubmitError> {
+    let meta;
+    unsafe {
+        meta = meta_data_create();
+    }
+    for (key, value) in meta_hm.iter() {
+        let c_key = CString::new(key.as_str()).map_err(|e| {
+            SubmitError::Field(
+                "meta key",
+                ArrayError::NullPresent(e.nul_position(), key.to_string()),
+            )
+        })?;
+        match value {
+            MetaValue::String(str) => {
+                let c_value = CString::new(str.as_str()).map_err(|e| {
+                    SubmitError::Field(
+                        "meta value",
+                        ArrayError::NullPresent(e.nul_position(), str.to_string()),
+                    )
+                })?;
+                unsafe {
+                    meta_data_add_string(meta, c_key.as_ptr(), c_value.as_ptr());
+                }
+            }
+            MetaValue::SignedInt(i) => unsafe {
+                meta_data_add_signed_int(meta, c_key.as_ptr(), *i);
+            },
+            MetaValue::UnsignedInt(u) => unsafe {
+                meta_data_add_unsigned_int(meta, c_key.as_ptr(), *u);
+            },
+            MetaValue::Double(d) => unsafe {
+                meta_data_add_double(meta, c_key.as_ptr(), *d);
+            },
+            MetaValue::Boolean(b) => unsafe {
+                meta_data_add_boolean(meta, c_key.as_ptr(), *b);
+            },
+        }
+    }
+    Ok(meta)
+}
+
+fn from_meta_data(
+    p: &str,
+    meta: *mut meta_data_t,
+) -> Result<Option<HashMap<String, MetaValue>>, ReceiveError> {
+    if meta.is_null() {
+        return Ok(None);
+    }
+
+    let mut toc: *mut *mut c_char = ptr::null_mut();
+    let count = unsafe { meta_data_toc(meta, &mut toc as *mut *mut *mut c_char) };
+    if count == 0 {
+        return Ok(None);
+    }
+    let mut meta_hm = HashMap::new();
+    // read meta and convert to hashmap entry
+    for i in 0..(count as isize) {
+        let c_key : &CStr;
```

Comment:
Considering all the individual unsafe expressions in this function, it may make sense to hoist unsafe a bit higher up in scope so that one doesn't feel forced to separate declaration and assignment.

Reply:
I try to keep  unsafe block small as necessary (but i don't know if it is really "safer")

I could change it like below to make it look better ?
```
let (c_key, key, value_type) = unsafe {
            let c_key : &CStr = CStr::from_ptr(*c_key_ptr);
            let key : String = c_key.to_str().map_err(|e| ReceiveError::Utf8(p.to_string(), "metadata key", e))?.to_string();
            let value_type : u32 = meta_data_type(meta, c_key.as_ptr()) as u32;
            (c_key, key, value_type)
        };
```